src/diag/diag_process.cpp

```diff
@@ -0,0 +1,344 @@
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the diagnostics module.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <common/code_utils.hpp>
+
+#include "diag_process.hpp"
+
+namespace Thread {
+
+const struct Command Diag::sCommands[] =
+{
+    { "start", &ProcessStart },
+    { "stop", &ProcessStop },
+    { "channel", &ProcessChannel },
+    { "power", &ProcessPower },
+    { "send", &ProcessSend },
+    { "repeat", &ProcessRepeat },
+    { "sleep", &ProcessSleep },
+    { "stats", &ProcessStats },
+};
+
+char Diag::sDiagOutput[MAX_DIAG_OUTPUT];
+struct DiagStats Diag::sStats;
+
+int Diag::sTxPower;
+unsigned int Diag::sChannel;
+unsigned int Diag::sTxPeriod;
+unsigned int Diag::sTxLen;
+unsigned int Diag::sTxPackets;
+
+void Diag::Init()
+{
+    sChannel = 20;
+    sTxPower = 0;
+    sTxPeriod = 0;
+    sTxLen = 0;
+    sTxPackets = 0;
+    memset(&sStats, 0, sizeof(struct DiagStats));
+}
+
+char *Diag::ProcessCmd(int argc, char *argv[])
+{
+    unsigned int i;
+
+    for (i = 0; i < sizeof(sCommands) / sizeof(sCommands[0]); i++)
+    {
+        if (strcmp(argv[0], sCommands[i].mName) == 0)
+        {
+            sCommands[i].mCommand(argc - 1, &argv[1], sDiagOutput);
+            break;
+        }
+    }
+
+    // more platform specific features will be processed under platform layer
+    if (i == sizeof(sCommands) / sizeof(sCommands[0]))
+    {
+        otPlatDiagProcess(argc, argv, sDiagOutput);
+    }
+
+    return sDiagOutput;
+}
+
+bool Diag::isEnabled()
+{
+    return otPlatDiagModeGet();
+}
+
+void Diag::ProcessStart(int argc, char *argv[], char *output)
+{
+    ThreadError error = kThreadError_None;
+
+    // enable radio
+    otPlatRadioEnable();
+
+    // enable promiscuous mode
+    otPlatRadioSetPromiscuous(true);
+
+    // stop timer
+    otPlatAlarmStop();
+
+    // start to listen on the default channel
+    SuccessOrExit(error = otPlatRadioReceive(sChannel));
+
+    // enable diagnostics mode
+    otPlatDiagModeSet(true);
+
+    memset(&sStats, 0, sizeof(struct DiagStats));
+
```

Comment:
how about putting line 122~ 126 to 'exit:' part and determine the output string according to error?
-- thus having only a single exit point